### PR TITLE
feat: Shared custom health check for multiple resources (#4212)

### PR DESCRIPTION
**Pull request by reggie-k**
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [V] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [V] Does this PR require documentation updates?
* [V] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [V] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [V] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by reggie-k** [comment]
#4212

**Comment by reggie-k** [comment]
@crenshaw-dev WDYT?

File: util/lua/lua.go [hunk 1]
```diff
@@ -333,6 +349,29 @@ func GetConfigMapKey(gvk schema.GroupVersionKind) string {
...
 }
 
+func GetWildcardConfigMapKey(vm VM, gvk schema.GroupVersionKind) (string, error) {
+	var gvkKeyToMatch string
+
+	if gvk.Group == "" {
+		gvkKeyToMatch = gvk.Kind
+	} else {
+		gvkKeyToMatch = fmt.Sprintf("%s/%s", gvk.Group, gvk.Kind)
+	}
```
Comment (crenshaw-dev) [inline util/lua/lua.go:359]:
```suggestion
	gvkKeyToMatch := GetGroupWildcardConfigMapKey(gvk)
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Just a couple tweaks to simplify! 

I merged master, so you'll need to pull this branch before making changes.

File: util/lua/lua.go (hunk 1 above)
Comment (reggie-k) [reply util/lua/lua.go]: Oh the GetGroupWildcardConfigMapKey was garbage left from initial attempts.
Deleted it. 
Removed code duplication using GetConfigMapKey.

File: util/lua/lua.go [hunk 2]
```diff
@@ -122,10 +124,24 @@ func (vm VM) ExecuteHealthLua(obj *unstructured.Unstructured, script string) (*h
...
 func (vm VM) GetHealthScript(obj *unstructured.Unstructured) (string, bool, error) {
+	// first, search the gvk as is in the ResourceOverrides
 	key := GetConfigMapKey(obj.GroupVersionKind())
+
 	if script, ok := vm.ResourceOverrides[key]; ok && script.HealthLua != "" {
 		return script.HealthLua, script.UseOpenLibs, nil
 	}
+
+	// if not found as is, perhaps it matches wildcard entries in the configmap
+	wildcardKey, err := GetWildcardConfigMapKey(vm, obj.GroupVersionKind())
```
[the remark this answers is not in the document]
Comment (reggie-k) [reply util/lua/lua.go]: Done

File: docs/operator-manual/health.md [hunk 3]
```diff
@@ -62,34 +64,50 @@ There are two ways to configure a custom health check. The next two sections des
 
 ### Way 1. Define a Custom Health Check in `argocd-cm` ConfigMap
 
-Custom health checks can be defined in `resource.customizations.health.<group_kind>` field of `argocd-cm`. If you are using argocd-operator, this is overridden by [the argocd-operator resourceCustomizations](https://argocd-operator.readthedocs.io/en/latest/reference/argocd/#resource-customizations).
+Custom health checks can be defined in
```
Comment (crenshaw-dev) [inline docs/operator-manual/health.md:67]: The yaml block breaking this up is strange to me. Is it intentional?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

**Review by crenshaw-dev — APPROVED** [review]
Thanks @reggie-k!

**Comment by Kerwood** [comment]
Did this feature get support for wildcards in the group also? Like,
```
      *.aws.crossplane.io/*:
        health.lua: |
           ...
```

**Comment by reggie-k** [comment]
Yeah, I will update the docs to state it clearly

**Comment by Kerwood** [comment]
> Yeah, I will update the docs to state it clearly

Awesome.. Thank you developing this feature :)

**Pull request merged**